chainntnfs/interface_test.go

```diff
@@ -1254,6 +1254,238 @@ func testCancelEpochNtfn(node *rpctest.Harness, notifier chainntnfs.TestChainNot
 	}
 }
 
+// testCatchUpOnMissedBlocks the case of multiple registered clients receiving
+// historical block epoch notifications due to the notifier's best known block
+// being out of date.
+func testCatchUpOnMissedBlocks(miner1 *rpctest.Harness,
+	notifier chainntnfs.TestChainNotifier, t *testing.T, notifierType string) {
+
+	const numBlocks = 10
+	const numClients = 5
+	var wg sync.WaitGroup
+	var extraBlockTest sync.WaitGroup
+
+	bestHash, bestHeight, err := miner1.Node.GetBestBlock()
+	if err != nil {
+		t.Fatalf("unable to get current blockheight %v", err)
+	}
+
+	// First, generate the blocks that the notifier will need to send
+	// historical notifications for.
+	_, err = miner1.Node.Generate(numBlocks)
+	if err != nil {
+		t.Fatalf("unable to generate blocks: %v", err)
+	}
+
+	// Next, start the notifier with outdated best block information.
+	outdatedBestBlock := chainntnfs.BlockEpoch{Height: bestHeight, Hash: bestHash}
+	if err := notifier.UnsafeStart(outdatedBestBlock); err != nil {
+		t.Fatalf("Unable to unsafe start the notifier: %v", err)
+	}
+
+	// Create numClients clients who will listen for block notifications.
+	// We expect each client to receive numBlocks + 1 notifications, 1 for each block
+	// that the notifier has missed out on.
+	// So we'll use a WaitGroup to synchronize the test.
+	for i := 0; i < numClients; i++ {
+		epochClient, err := notifier.RegisterBlockEpochNtfn(nil)
+		if err != nil {
+			t.Fatalf("unable to register for epoch notification: %v", err)
+		}
+
+		// We use numBlocks + 1 because we generate numBlocks then
+		// 1 additional block to trigger the historical notifications.
+		wg.Add(numBlocks + 1)
+		extraBlockTest.Add(1)
+		go func() {
+			defer extraBlockTest.Done()
+
+			expectedHeight := outdatedBestBlock.Height + 1
+			for i := 0; i < numBlocks+1; i++ {
+				block := <-epochClient.Epochs
+				if block.Height != expectedHeight {
+					t.Fatalf("received block of height: %d, "+
+						"expected: %d", block.Height,
+						expectedHeight)
+				}
+				expectedHeight++
+				wg.Done()
+			}
+
+			// Ensure that we don't receive an unexpected additional
+			// notification.
+			select {
+			case <-epochClient.Epochs:
+				t.Fatalf("received unexpected block notification")
+			case <-time.After(5 * time.Second):
+			}
+		}()
+	}
+
+	// Generate a single block to trigger the backlog of historical
+	// notifications for the previously mined blocks.
+	// Each client should receive numBlocks + 1 notifications, thereby
+	// unblocking the goroutine above.
+	if _, err := miner1.Node.Generate(1); err != nil {
```

Comment:
Same comment about simplification. After mining this additional block, you could do the serial loop that goes through all clients, checking the received block epochs.

Reply:
fixed